code/game/objects/items/devices/lighting/flashlight.dm

```diff
@@ -15,29 +16,105 @@
 	uv_intensity = 50
 	light_wedge = LIGHT_WIDE
 
-	matter = list(DEFAULT_WALL_MATERIAL = 50, MATERIAL_GLASS = 20)
+	matter = list(MATERIAL_PLASTIC = 50, MATERIAL_GLASS = 20)
 
 	action_button_name = "Toggle Flashlight"
-	var/on = 0
-	var/brightness_on = 3 //luminosity when on
+	/// Is the light currently on or off?
+	var/on = FALSE
+	/// Luminosity when on
+	var/brightness_on = 4
+	// Lighting power when on
+	var/flashlight_power = 0.8
+	/// Sound the light makes when it's switched
+	var/toggle_sound = 'sound/machines/switch_flashlight.ogg'
+	/// Sound the light makes when it's turned on
 	var/activation_sound = 'sound/items/flashlight.ogg'
-	var/spawn_dir // a way for mappers to force which way a flashlight faces upon spawning
+	var/obj/item/cell/cell
+	var/cell_type = /obj/item/cell/device
+	var/list/brightness_levels
+	var/brightness_level = "medium"
+	var/power_usage
+	/// Does the light use power?
+	var/power_use = TRUE
+	/// Should the item start with a cell?
+	var/starts_with_cell = TRUE
+	/// Does the item accept large cells?
+	var/accepts_large_cells = FALSE
+	/// Is the light always on?
+	var/always_on = FALSE
+	/// How efficient the flashlight is at producing light compared to baseline
+	var/efficiency_modifier = 1.0
+	/// A way for mappers to force which way a flashlight faces upon spawning
+	var/spawn_dir
 
 /obj/item/device/flashlight/Initialize()
+
+	if(power_use && cell_type)
+		if(starts_with_cell)
+			cell = new cell_type(src)
+		brightness_levels = list("low" = 0.25, "medium" = 0.5, "high" = 1)
+		power_usage = (brightness_levels[brightness_level] / efficiency_modifier)
+	else
+		verbs -= /obj/item/device/flashlight/verb/toggle_brightness
+
 	if (on)
-		light_range = brightness_on
+		if(brightness_level == "low")
+			light_range = brightness_on * 0.5
+		else if(brightness_level == "high")
+			light_range = brightness_on * 1.5
+		else
+			light_range = brightness_on
 		update_icon()
+
 	. = ..()
+
 	if(light_wedge)
 		set_dir(pick(NORTH, SOUTH, EAST, WEST))
 		if(spawn_dir)
 			set_dir(spawn_dir)
 		update_light()
 
+/obj/item/device/flashlight/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	QDEL_NULL(cell)
+	return ..()
+
+/obj/item/device/flashlight/get_cell()
+	return cell
+
+/obj/item/device/flashlight/proc/set_brightness(mob/user)
+	var/choice = input("Choose a brightness level.") as null|anything in brightness_levels
+	if(choice)
+		brightness_level = choice
+		power_usage = brightness_levels[choice]
+		to_chat(user, SPAN_NOTICE("You set the brightness level on \the [src] to [brightness_level]."))
+		update_icon()
+
+/obj/item/device/flashlight/process()
```

Comment:
```suggestion
/obj/item/device/flashlight/process()
	if((on && cell && brightness_level && power_usage)
		visible_message(
			SPAN_WARNING("\The [src] flickers before shutting out!")
		)
		playsound(src.loc, 'sound/effects/sparks3.ogg', 10, 1, -3)
		toggle()
	return PROCESS_KILL
```

Reply:
The flickering is not represented in-scene. It just turns off the light completely.

Reply:
Perfect, I playtested it and can confirm, feel free to dismiss this suggestion